WL4EditorWindow.cpp

```diff
@@ -1880,3 +1904,52 @@ void WL4EditorWindow::on_actionRect_Select_Mode_toggled(bool arg1)
 {
     ui->graphicsView->SetRectSelectMode(arg1);
 }
+
+/// <summary>
+/// Add a new Room to the current Level.
+/// </summary>
+void WL4EditorWindow::on_actionNew_Room_triggered()
+{
+    // Create new Room based on current Room
+    int newRoomId = CurrentLevel->GetRooms().size();
+    if (newRoomId == 16)
+    {
+        OutputWidget->PrintString("Cannot add more Room to the current Level!");
+    }
+    int roomTableAddress = ROMUtils::PointerFromData(WL4Constants::RoomDataTable + CurrentLevel->GetLevelID() * 4);
+    CurrentLevel->AddRoom(new LevelComponents::Room(roomTableAddress + selectedRoom * 0x2C, newRoomId, CurrentLevel->GetLevelID()));
+    OutputWidget->PrintString("Create new Room based on current Room.");
+
+    // Add one Door to the new Room as well as spriteset settings
+    {
+        // Create a new door struct with blank fields
+        LevelComponents::__DoorEntry newDoorEntry;
+        memset(&newDoorEntry, 0, sizeof(LevelComponents::__DoorEntry));
+
+        // Initialize the fields
+        newDoorEntry.DoorTypeByte = (unsigned char) 2;
+        newDoorEntry.EntitySetID = (unsigned char) 1;
+        newDoorEntry.RoomID = (unsigned char) newRoomId;
+        newDoorEntry.DoorTypeByte = LevelComponents::DoorType::Instant;
+        LevelComponents::Door *newDoor =
+            new LevelComponents::Door(newDoorEntry, (unsigned char) newRoomId, CurrentLevel->GetDoors().size());
+        int entitysetId = CurrentLevel->GetRooms()[selectedRoom]->GetCurrentEntitySetID();
+        newDoor->SetEntitySetID(entitysetId);
+        newDoor->SetDestinationDoor(CurrentLevel->GetDoors()[0]);
+
+        // Add the new door to the Level object and re-render the screen
+        CurrentLevel->AddDoor(newDoor);
+
+        // Set Current Entity list
+        CurrentLevel->GetRooms()[newRoomId]->SetCurrentEntitySet(entitysetId);
+    }
+
+    // Reset LevelHeader param
+    CurrentLevel->GetLevelHeader()->NumOfMap++;
+
+    // UI updates
+    SetCurrentRoomId(newRoomId);
+
+    // Clear everything in the new room
+    ClearEverythingInRoom(true, selectedRoom);
```

Comment:
I'm confused by the logging statement above. Is a blank room created or is a copy created based on the previous room?

Reply:
it uses the current room's settings, all the settings, then we don't need to care about things like entityset id which stored in the current door data, and also the Tileset id stored in the current RoomHeader, when creating the new Room, and the code is easy to write too.

Reply:
But is the new room blank?

Reply:
If it is, we should change the logging statement to "Created a new blank room (# newRoomId) using the current room's settings."

Reply:
yes, it is blank, Done.